File: matlab/computegigexpectations.m
function [Ex Exinv] = computegigexpectations(alpha, beta, gamma)
% function [Ex Exinv] = computegigexpectations(alpha, beta, gamma)

if (numel(alpha) == 1)
    alpha = alpha * ones(size(beta));
end

Ex = zeros(size(beta));
Exinv = zeros(size(beta));

% For very small values of gamma and positive values of alpha, the GIG
% distribution becomes a gamma distribution, and its expectations are both
% cheaper and more stable to compute that way.
giginds = find(gamma(:) > 1e-200);
gaminds = find(gamma(:) <= 1e-200);
if (sum(alpha(gaminds) < 0) > 0)
    error('problem with arguments.');
end

% Compute expectations for GIG distribution.
sqrtbeta = sqrt(beta(giginds));
sqrtgamma = sqrt(gamma(giginds));
% besselparam = sqrtbeta .* sqrtgamma;
% % Note that we're using the *scaled* version here, since we're just
% % computing ratios and it's more stable.
besselalphaminus = besselk(alpha(giginds)-1, 2*sqrtbeta.*sqrtgamma, 1);
besselalpha = besselk(alpha(giginds), 2*sqrtbeta.*sqrtgamma, 1);
besselalphaplus = besselk(alpha(giginds)+1, 2*sqrtbeta.*sqrtgamma, 1);
sqrtratio = sqrtgamma ./ sqrtbeta;
Ex(giginds) = besselalphaplus .* sqrtratio ./ besselalpha;
Exinv(giginds) = besselalphaminus ./ (sqrtratio .* besselalpha);

% Compute expectations for gamma distribution where we can get away with 
% it.
Ex(gaminds) = alpha(gaminds) ./ beta(gaminds);
Exinv(gaminds) = beta(gaminds) ./ (alpha(gaminds) - 1);
Exinv(Exinv < 0) = inf;

File: R/computegigexpectations.R
computegigexpectations <- function(alpha, beta, gamma) {
    if (length(alpha) == 1) {
        alpha = alpha * rep(1, length(beta))
    }

    Ex <- rep(0, length(beta))
    Exinv <- rep(0, length(beta))

    giginds <- which(gamma > 1e-200)
    gaminds <- which(gamma > 1e-200)

    if (sum(alpha(gaminds) < 0) > 0) {
        stop("problem with arguments.")
    }

    sqrtbeta <- sqrt(beta[giginds])
    sqrtgamma <- sqrt(gamma[giginds])

    besselalphaminus <- besselK(2 * sqrtbeta * sqrtgamma, alpha[giginds]-1, TRUE)
    besselalpha <- besselK(2 * sqrtbeta * sqrtgamma, alpha[giginds], TRUE)
    besselalphaplus <- besselK(2 * sqrtbeta * sqrtgamma, alpha[giginds]+1, TRUE)

    sqrtratio <- sqrtgamma / sqrtbeta
    Ex[giginds] <- besselalphaplus * sqrtratio / besselalpha
    Exinv[giginds] <- besselalphaminus / (sqrtratio * besselalpha)

    Ex[gaminds] <- alpha[gaminds] / beta[gaminds]
    Exinv[gaminds] <- beta[gaminds] / (alpha[gaminds] - 1)
    Exinv[Exinv < 0] <- Inf

    return(list(Ex=Ex, Exinv=Exinv))
}
